File: ios/Classes/FrameAnalyzer.swift
import CoreMedia
import ImageIO
import Vision

/// The decoding hot path on iOS. Invoked per video frame on the capture queue.
///
///  - **skips frames** to honor the FPS cap (preview is unaffected);
///  - reuses a single `VNDetectBarcodesRequest` (no per-frame allocation);
///  - filters detections to the configured **scan area**;
///  - applies the **duplicate filter** and honors the **scan mode**.
///
/// Results are emitted via [onBarcodes] as already-serialized dictionaries.
final class FrameAnalyzer {
  private let config: ScannerConfig
  private let request: VNDetectBarcodesRequest
  private let dupFilter: DuplicateFilter
  private let onBarcodes: ([[String: Any?]]) -> Void
  private let onError: (String) -> Void

  private var lastAnalyzed: CFTimeInterval = 0

  /// Gate toggled by start/pause/stop.
  var active: Bool = false

  /// Orientation used to present frames upright to Vision. Updated by the
  /// camera manager on device-orientation changes.
  var orientation: CGImagePropertyOrientation = .right

  init(
    config: ScannerConfig,
    onBarcodes: @escaping ([[String: Any?]]) -> Void,
    onError: @escaping (String) -> Void
  ) {
    self.config = config
    self.dupFilter = DuplicateFilter(timeoutMs: config.duplicateTimeoutMs)
    self.onBarcodes = onBarcodes
    self.onError = onError
    self.request = VNDetectBarcodesRequest()
    self.request.symbologies = FormatMapper.toSymbologies(config.formatMask)
    if #available(iOS 17.0, *) {
      // Hint Vision toward low-latency single-frame decoding.
      self.request.revision = VNDetectBarcodesRequestRevision3
    }
  }

  func resetDuplicates() { dupFilter.reset() }

  func analyze(_ sampleBuffer: CMSampleBuffer) {
    guard active else { return }
    let now = CACurrentMediaTime()
    if now - lastAnalyzed < config.frameInterval { return }  // frame skip
    lastAnalyzed = now

    guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else { return }
    let width = CVPixelBufferGetWidth(pixelBuffer)
    let height = CVPixelBufferGetHeight(pixelBuffer)
    let oriented = CoordinateMapper.orientedSize(
      width: width, height: height, orientation: orientation)

    let handler = VNImageRequestHandler(
      cvPixelBuffer: pixelBuffer, orientation: orientation, options: [:])
    do {
      try handler.perform([request])
    } catch {
      onError(error.localizedDescription)
      return
    }
    guard let results = request.results, !results.isEmpty else { return }
    handle(results, orientedSize: oriented, now: now)
  }

  private func handle(
    _ results: [VNBarcodeObservation], orientedSize: CGSize, now: CFTimeInterval
  ) {
    guard active else { return }

    let inArea = results.filter {
      CoordinateMapper.isInside(box: $0.boundingBox, area: config.scanArea)
    }
    let fresh = inArea.filter {
      dupFilter.shouldEmit(BarcodeMapper.dedupKey($0), now: now)
    }
    if fresh.isEmpty { return }

    let epochMs = Int64(Date().timeIntervalSince1970 * 1000)
    func map(_ o: VNBarcodeObservation) -> [String: Any?] {
      BarcodeMapper.toMap(o, orientedSize: orientedSize, timestampMs: epochMs)
    }

    switch config.scanMode {
    case 0:  // single
      active = false
      if let first = fresh.first { onBarcodes([map(first)]) }
    case 2:  // multi
      onBarcodes(fresh.map(map))
    default:  // continuous
      if !config.continuousScanning { active = false }
      onBarcodes(fresh.map(map))
    }
  }
}

File: android/src/main/kotlin/com/umda/barcode_scanner_pro/FrameAnalyzer.kt
package com.umda.barcode_scanner_pro

import android.os.SystemClock
import androidx.camera.core.ExperimentalGetImage
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import com.google.mlkit.vision.barcode.BarcodeScanner
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.common.InputImage

/**
 * The decoding hot path. Runs on a background executor (see [CameraManager]) and:
 *
 *  - **skips frames** so decoding never exceeds the configured FPS cap, while the
 *    preview keeps running at the camera's native rate;
 *  - reuses a single [BarcodeScanner] instance (no per-frame allocation of the
 *    detector);
 *  - filters detections to the configured **scan area**;
 *  - applies the **duplicate filter**;
 *  - honors **scan mode** (single / continuous / multi).
 *
 * Results are handed back through [onBarcodes] as already-serialized maps so the
 * platform view only has to forward them to the event channel.
 */
internal class FrameAnalyzer(
    private val scanner: BarcodeScanner,
    private val config: ScannerConfig,
    private val onBarcodes: (List<Map<String, Any?>>) -> Unit,
    private val onError: (String) -> Unit,
) : ImageAnalysis.Analyzer {

    private val dupFilter = DuplicateFilter(config.duplicateTimeoutMs)
    private var lastAnalyzedMs = 0L

    /** Gate toggled by start/pause/stop. When false, frames are dropped cheaply. */
    @Volatile
    var active: Boolean = false

    fun resetDuplicates() = dupFilter.reset()

    @ExperimentalGetImage
    override fun analyze(imageProxy: ImageProxy) {
        if (!active) {
            imageProxy.close()
            return
        }
        val now = SystemClock.elapsedRealtime()
        if (now - lastAnalyzedMs < config.frameIntervalMs) {
            imageProxy.close() // frame skip
            return
        }
        lastAnalyzedMs = now

        val mediaImage = imageProxy.image
        if (mediaImage == null) {
            imageProxy.close()
            return
        }
        val rotation = imageProxy.imageInfo.rotationDegrees
        val input = InputImage.fromMediaImage(mediaImage, rotation)
        val width = mediaImage.width
        val height = mediaImage.height

        scanner.process(input)
            .addOnSuccessListener { barcodes ->
                handle(barcodes, width, height, rotation, now)
            }
            .addOnFailureListener { e ->
                onError(e.localizedMessage ?: "Barcode decoding failed")
            }
            .addOnCompleteListener {
                imageProxy.close() // release buffer back to the pool
            }
    }

    private fun handle(
        barcodes: List<Barcode>,
        width: Int,
        height: Int,
        rotation: Int,
        nowMs: Long,
    ) {
        if (!active || barcodes.isEmpty()) return

        val inArea = barcodes.filter {
            val box = it.boundingBox ?: return@filter false
            CoordinateMapper.isInside(box, config.scanArea, width, height, rotation)
        }
        if (inArea.isEmpty()) return

        val fresh = inArea.filter {
            dupFilter.shouldEmit(BarcodeMapper.dedupKey(it), nowMs)
        }
        if (fresh.isEmpty()) return

        val epochMs = System.currentTimeMillis()
        when (config.scanMode) {
            // Single: emit only the first and stop decoding.
            0 -> {
                active = false
                onBarcodes(
                    listOf(BarcodeMapper.toMap(fresh.first(), width, height, rotation, epochMs)),
                )
            }
            // MultiBarcode: emit the whole batch at once.
            2 -> onBarcodes(
                fresh.map { BarcodeMapper.toMap(it, width, height, rotation, epochMs) },
            )
            // Continuous (default).
            else -> {
                if (!config.continuousScanning) active = false
                onBarcodes(
                    fresh.map { BarcodeMapper.toMap(it, width, height, rotation, epochMs) },
                )
            }
        }
    }
}
